stablehlo/dialect/Base.cpp

```diff
@@ -261,67 +267,114 @@ FailureOr<std::pair<int64_t, int64_t>> inferMergedDimAndBound(
   return std::make_pair(size, bound);
 }
 
-// TODO(zhouxin) Refactor to better handle errors and return single type
-LogicalResult inferMostSpecificType(
-    Optional<Location> location, TypeRange inputTypes,
-    SmallVectorImpl<Type>& inferredReturnTypes) {
-  SmallVector<RankedTensorType> rankedTypes;
-  for (auto inputType : inputTypes)
-    if (auto rankedType = inputType.dyn_cast<RankedTensorType>())
-      rankedTypes.push_back(rankedType);
-  if (rankedTypes.empty()) {
-    inferredReturnTypes.push_back(inputTypes[0]);
-    return success();
+// Inference rules for conditional branches (lhs/rhs are commutative):
+//       Dim of lhs     Dim of rhs      Infer
+//  c0:  X              X               X
+//  c1:  X              ?               ?
+//  c2:  X              ?, B            ?, max(X, B)
+//  c3:  ?              ?               ?
+//  c4:  ?              ?, B            ?
+//  c5:  ?, B           ?, C            ?, max(B, C)
+FailureOr<std::pair<int64_t, int64_t>> inferLeastSpecificDimAndBound(
+    Optional<Location> location, int64_t dim, int64_t leftSize,
+    int64_t rightSize, int64_t leftBound, int64_t rightBound) {
+  bool isLeftStaticDim = !isDynamicDimSize(leftSize);
+  bool isRightStaticDim = !isDynamicDimSize(rightSize);
+  bool isLeftStaticBound = !isDynamicDimSize(leftBound);
+  bool isRightStaticBound = !isDynamicDimSize(rightBound);
+  int64_t inferredSize = ShapedType::kDynamic;
+  int64_t inferredBound = ShapedType::kDynamic;
+
+  if (isLeftStaticDim || isRightStaticDim) {
+    if (isLeftStaticDim && isRightStaticDim) {
+      if (leftSize != rightSize)
+        return emitOptionalError(location, "Mismatched dimension sizes ",
+                                 leftSize, " and ", rightSize, " in dimension ",
+                                 dim);
+      inferredSize = leftSize;
+    } else if (isLeftStaticBound || isRightStaticBound) {
+      inferredBound = isLeftStaticDim ? std::max(leftSize, rightBound)
+                                      : std::max(rightSize, leftBound);
+    }
+  } else if (isLeftStaticBound && isRightStaticBound) {
+    inferredBound = std::max(leftBound, rightBound);
   }
+  return std::make_pair(inferredSize, inferredBound);
+}
 
+FailureOr<TensorType> inferTypeWithCustomFunc(
+    Optional<Location> location, SmallVector<RankedTensorType> rankedTypes,
+    std::function<FailureOr<std::pair<int64_t, int64_t>>(
+        Optional<Location>, int64_t, int64_t, int64_t, int64_t, int64_t)>
+        inferDimAndBoundFunc) {
```

Comment:
How about simplifying this to `inferDimAndBound`?

Reply:
I prefer to simplify it to inferDimAndBoundFn, I have not find a code style for this in LLVM, but I think it provides a clear clue that this is a fn to be used, like https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Transforms/DialectConversion.h#L237